### PR TITLE
Modified BigQueryDynamicDestination destination to support write API with exactly once semantics

### PR DESCRIPTION
**Pull request by dedocibula**
Spanner change stream to BQ template fails to launch with error "ParDo requires a deterministic key coder in order to use state and timers" when using `useStorageWriteApi=true` and `useStorageWriteApiAtLeastOnce=false` with:

```
com.google.cloud.teleport.v2.common.UncaughtExceptionLogger - The template launch failed.
java.lang.IllegalArgumentException: ParDo requires a deterministic key coder in order to use state and timers, the reason is:
 ShardedKey$Coder(KvCoder(SerializableCoder(com.google.cloud.bigquery.TableId),TableRowJsonCoder)) is not deterministic because:
	Key coder must be deterministic
	at org.apache.beam.sdk.transforms.ParDo.validateStateApplicableForInput(ParDo.java:448)
	at org.apache.beam.sdk.transforms.ParDo.access$300(ParDo.java:394)
	at org.apache.beam.sdk.transforms.ParDo$MultiOutput.expand(ParDo.java:953)
	at org.apache.beam.sdk.transforms.ParDo$MultiOutput.expand(ParDo.java:872)
	at org.apache.beam.sdk.Pipeline.applyInternal(Pipeline.java:559)
	at org.apache.beam.sdk.Pipeline.applyTransform(Pipeline.java:491)
	at org.apache.beam.sdk.values.PCollection.apply(PCollection.java:352)
	at org.apache.beam.sdk.transforms.ParDo$SingleOutput.expand(ParDo.java:797)
	at org.apache.beam.sdk.transforms.ParDo$SingleOutput.expand(ParDo.java:712)
	at org.apache.beam.sdk.Pipeline.applyInternal(Pipeline.java:559)
	at org.apache.beam.sdk.Pipeline.applyTransform(Pipeline.java:491)
	at org.apache.beam.sdk.values.PCollection.apply(PCollection.java:352)
	at org.apache.beam.sdk.transforms.GroupIntoBatches.expand(GroupIntoBatches.java:338)
	at org.apache.beam.sdk.transforms.GroupIntoBatches.expand(GroupIntoBatches.java:103)
	at org.apache.beam.sdk.Pipeline.applyInternal(Pipeline.java:559)
	at org.apache.beam.sdk.Pipeline.applyTransform(Pipeline.java:508)
	at org.apache.beam.sdk.values.PCollection.apply(PCollection.java:365)
	at org.apache.beam.sdk.io.gcp.bigquery.StorageApiLoads.expandTriggered(StorageApiLoads.java:231)
	at org.apache.beam.sdk.io.gcp.bigquery.StorageApiLoads.expand(StorageApiLoads.java:125)
	at org.apache.beam.sdk.io.gcp.bigquery.StorageApiLoads.expand(StorageApiLoads.java:46)
	at org.apache.beam.sdk.Pipeline.applyInternal(Pipeline.java:559)
	at org.apache.beam.sdk.Pipeline.applyTransform(Pipeline.java:508)
	at org.apache.beam.sdk.values.PCollection.apply(PCollection.java:365)
	at org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO$Write.continueExpandTyped(BigQueryIO.java:3718)
	at org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO$Write.expandTyped(BigQueryIO.java:3502)
	at org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO$Write.expand(BigQueryIO.java:3374)
	at org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO$Write.expand(BigQueryIO.java:2214)
	at org.apache.beam.sdk.Pipeline.applyInternal(Pipeline.java:559)
	at org.apache.beam.sdk.Pipeline.applyTransform(Pipeline.java:508)
	at org.apache.beam.sdk.values.PCollection.apply(PCollection.java:365)
	at com.google.cloud.teleport.v2.templates.spannerchangestreamstobigquery.SpannerChangeStreamsToBigQuery.run(SpannerChangeStreamsToBigQuery.java:382)
	at com.google.cloud.teleport.v2.templates.spannerchangestreamstobigquery.SpannerChangeStreamsToBigQuery.main(SpannerChangeStreamsToBigQuery.java:198)
```

Given that `TableRowJsonCoder`'s verifyDeterministic method always throws we need to implement a custom deterministic coder for the value

### CODE REVIEW

File: v2/googlecloud-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/templates/spannerchangestreamstobigquery/BigQueryDynamicDestinations.java [hunk 1]
```diff
@@ -68,35 +78,34 @@ private BigQueryDynamicDestinations(
...
 
-    return TableId.of(bigQueryProject, bigQueryDataset, bigQueryTableName);
+    return String.format("%s:%s.%s", bigQueryProject, bigQueryDataset, bigQueryTableName);
   }
 
   @Override
-  public KV<TableId, TableRow> getDestination(ValueInSingleWindow<TableRow> element) {
-    TableRow tableRow = element.getValue();
-    return KV.of(getTableId(bigQueryTableTemplate, tableRow), tableRow);
+  public Coder<KV<String, List<TableFieldSchema>>> getDestinationCoder() {
```
Comment (damondouglas) [inline v2/googlecloud-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/templates/spannerchangestreamstobigquery/BigQueryDynamicDestinations.java:89]: May we trouble you to add a test for whether the `Coder<KV<String, List<TableFieldSchema>>>` decode/encode yeild matching results, particularly in the setting of a `RECORD` `getType()` field where said RECORD has itself multiple fields?

**Review by damondouglas — CHANGES_REQUESTED** [review]

File: v2/googlecloud-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/templates/spannerchangestreamstobigquery/BigQueryDynamicDestinations.java (hunk 1 above)
Comment (damondouglas) [reply v2/googlecloud-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/templates/spannerchangestreamstobigquery/BigQueryDynamicDestinations.java:89]: The `codecov/patch` failure among the checks reminded me of this.

Comment (dedocibula) [reply v2/googlecloud-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/templates/spannerchangestreamstobigquery/BigQueryDynamicDestinations.java:89]: Added unit test for getDestinationCoder

**Review by damondouglas — APPROVED** [review]
Approved and  merging:
- Java PR / JUnit Test Report (pull_request)  failed checks have nothing to do with this PR
- Dataflow Integration Test failures show the same failures in the JUnit Test Report

**Pull request merged**